### PR TITLE
Fix temporary inventory in `InvRef:set_lists()` and type check `listname`

### PR DESCRIPTION
**Pull request by OgelGames**
Currently when calling `InvRef:set_lists()` with a bad value (e.g. `{main=true}`), all lists are cleared. This PR fixes that by fixing the creation of a temporary inventory, so a new inventory is created instead of a reference. It also adds a missing type check to `listname`.

## How to test

Run this code:
```lua
local inv = minetest.create_detached_inventory("test")
inv:set_lists({main = {"air", ""})
print("Before error: ", dump(inv:get_lists()))
print(pcall(inv.set_lists, inv, {main=true}))
print("After error: ", dump(inv:get_lists()))
```
Without the fix in this PR, the output should be: (inventory empty after error)
```
Before error:   {
        main = {
                ItemStack("air"),
                ItemStack(""),
        }
}
false   bad argument #4 to '?' (table expected, got boolean)
After error:    {

}
```
With the fix, the output should be: (no change after error)
```
Before error:   {
        main = {
                ItemStack("air"),
                ItemStack(""),
        }
}
false   bad argument #4 to '?' (table expected, got boolean)
After error:    {
        main = {
                ItemStack("air"),
                ItemStack(""),
        }
}
```

### CODE REVIEW

File: src/script/lua_api/l_inventory.cpp [hunk 1]
```diff
@@ -264,19 +264,19 @@ int InvRef::l_set_lists(lua_State *L)
...
 	luaL_checktype(L, 2, LUA_TTABLE);
 	while (lua_next(L, 2)) {
-		const char *listname = lua_tostring(L, -2);
-		read_inventory_list(L, -1, tempInv, listname, server);
+		const char *listname = luaL_checkstring(L, -2);
+		read_inventory_list(L, -1, &tempInv, listname, server);
 		lua_pop(L, 1);
 	}
-	inv = tempInv;
+	*inv = tempInv;
```
Comment (sfan5) [inline src/script/lua_api/l_inventory.cpp:279]:
```suggestion
	*inv = std::move(tempInv);
```

**Review by sfan5 — COMMENTED** [review]
Also consider adding an unit test [here](https://github.com/minetest/minetest/tree/master/games/devtest/mods/unittests).

File: src/script/lua_api/l_inventory.cpp (hunk 1 above)
Comment (OgelGames) [reply src/script/lua_api/l_inventory.cpp:279]: I might not be understanding this correctly, but I don't think that is needed; the assignment operator is overridden here, which copies the inventory: https://github.com/minetest/minetest/blob/c63c05b141d994e801c82344499d86f5b67a2e3f/src/inventory.cpp#L854-L867

**Comment by OgelGames** [comment]
> Also consider adding an unit test here

I'll do that in another PR 👍

File: src/script/lua_api/l_inventory.cpp (hunk 1 above)
Comment (sfan5) [reply src/script/lua_api/l_inventory.cpp:279]: This enables the inventory to be moved not copied. Might not work in this case if there's a custom copy-operator.

**Comment by sfan5** [comment]
> I'll do that in another PR 👍

I think it makes more sense to keep a fix and the relevant test in the same PR.

**Comment by grorp** [comment]
Considering the amount of inventory-related regressions, wouldn't it make sense to even require a unittest for any change to the inventory code?

**Comment by OgelGames** [comment]
> I think it makes more sense to keep a fix and the relevant test in the same PR.

Should I combine this PR and the unit tests with #14566, and have everything in one PR?

**Comment by sfan5** [comment]
> Should I combine this PR and the unit tests with #14566, and have everything in one PR?

If you want to, sure that's okay.

**Pull request closed without merging**